build_tools/bazel/iree_trace_runner_test.bzl

```diff
@@ -6,7 +6,222 @@
 
 """Macros for defining tests that use a trace-runner."""
 
-def iree_generated_trace_runner_test(**kwargs):
-    # TODO: implement this. For now, it's only parsed by bazel_to_cmake.py, so
-    # the iree_generated_check_test's are just omitted in bazel builds.
-    pass
+load("//iree/tools:compilation.bzl", "iree_bytecode_module")
+load("//build_tools/bazel:run_binary_test.bzl", "run_binary_test")
+
+def iree_trace_runner_test(
+        name,
+        src,
+        module,
+        target_backend,
+        driver,
+        trace_runner,
+        trace,
+        compiler_flags = [],
+        runner_args = [],
+        opt_tool = "//iree/tools:iree-opt",
+        opt_flags = [],
+        tags = [],
+        timeout = None,
+        **kwargs):
+    """Creates a test running a custom trace-runner on a trace file (yaml).
+
+    Args:
+        name: Name of the target
+        src: mlir source file to be compiled to an IREE module.
+        target_backend: target backend to compile for.
+        driver: driver to run the module with.
+        compiler_flags: additional flags to pass to the compiler. Bytecode
+            translation and backend flags are passed automatically.
+        runner_args: additional args to pass to the trace-runner program. The driver
+            and input file flags are passed automatically.
+        tags: Additional labels to apply to the test. The package path and
+            "driver=${DRIVER}" are added automatically.
+        opt_tool: Defaulting to iree-opt. Tool used to preprocess the source files
+            if opt_flags is specified.
+        opt_flags: If specified, source files are preprocessed with opt_tool with
+            these flags.
+        trace_runner: trace-runner program to run.
+        trace: trace file input to the trace-runner program.
+        module: specifies the  path to use for the enerated IREE module (.vmfb). Mandatory,
+            unlike in iree_check_test, because trace files (.yaml) reference a specific module file path.
+        timeout: timeout for the generated tests.
+        **kwargs: any additional attributes to pass to the underlying tests and test suite.
+    """
+
+    bytecode_module_name = name + "_bytecode_module"
+    iree_bytecode_module(
+        name = bytecode_module_name,
+        module = module,
+        src = src,
+        flags = [
+            "-iree-mlir-to-vm-bytecode-module",
+            "-mlir-print-op-on-diagnostic=false",
+            "-iree-hal-target-backends=%s" % target_backend,
+        ] + compiler_flags,
+        opt_tool = opt_tool,
+        opt_flags = opt_flags,
+        visibility = ["//visibility:private"],
```

Comment:
This should probably get kwargs

Reply:
Ah good point! But, then, I'm removing kwards from the below `run_binary_test`. Otherwise, kwargs would only be usable for the intersection of those args that are accepted by both `iree_bytecode_module` and `run_binary_test` (and with the "same" semantics in some sense). For the same reason, below, I'm removing kwards from `test_suite`.

Reply:
Ah no, we should be using kwargs for exactly the case of kwargs that apply to all targets. Things like testonly, tags, etc. If there's something that needs to be passed and only applies to some targets then we need to define it as an explicit kwarg and clarify what it is passed to.

Reply:
ah ok, thanks for the explanation!